src/imputation/tmi_imputation.py

```diff
@@ -0,0 +1,337 @@
+
+    # Filter for q604 = No or status = "Form sent out"
+    filtered_data = long_df.loc[
+        (long_df["status"] == "Form sent out") | (long_df["604"] == "No")
+    ]
+
+    filtered_data = sic_to_pg_mapper(filtered_data, sic_mapper, target_col="201", formtype = "0001")
+
+    updated_df = apply_to_original(filtered_data, df)
+
+    return updated_df
+
+
+# # Impute Business R&D type TODO: STEP 2
+# filtered_data["200"] = np.random.choice(
+#     ["C", "D"], size=len(filtered_data), p=[0.7, 0.3]
+# )
+# cur_df = apply_to_original(filtered_data, cur_df)
+
+
+def create_imp_class_col(
+    clean_df: pd.DataFrame, col_first_half: str, col_second_half: str, class_name: str
+) -> pd.DataFrame:
+
+    # Create class col with concatenation
+    clean_df[f"{class_name}"] = (
+        clean_df[f"{col_first_half}"].astype(str)
+        + "_"
+        + clean_df[f"{col_second_half}"].astype(str)
+    )
+
+    fil_df = filter_by_column_content(clean_df, "cellnumber", [817])
+    # Create class col with concatenation + 817
+    fil_df[f"{class_name}"] = fil_df[f"{class_name}"] + "_817"
+
+    clean_df = apply_to_original(fil_df, clean_df)
+
+    return clean_df
+
+
+def apply_filters(df):
+
+    clean_statuses = ["Clear", "Clear - overridden", "Clear - overridden SE"]
+    filtered_df = filter_by_column_content(df, "status", clean_statuses)
+
+    return filtered_df
+
+
+def fill_zeros(df: pd.DataFrame, column: str):
+    return df[column].replace({math.nan: 0}).astype("float")
+
+
+def clean_no_rd_data(filtered_df, og_df, target_variables: list):
+    """TODO: Check if we need to do this"""
+    # Replace 305 nulls with zeros
+    filtered_df["305"] = fill_zeros(filtered_df, "305")
+    og_df = apply_to_original(filtered_df, og_df)
+
+    # Replace Nan with zero for companies with NO R&D Q604 = "No"
+    no_rd = filter_by_column_content(og_df, "604", ["No"])
+
+    for i in target_variables:
+        no_rd[i] = fill_zeros(no_rd, i)
+
+    og_df = apply_to_original(no_rd, og_df)
+
+    # Return cleaned original dataset
+    return og_df
+
+
+def tmi_pre_processing(df, target_variables_list: list) -> pd.DataFrame:
+
+    # Create a copy for reference
+    copy_df = df.copy()
+
+    # Filter for long form data
+    long_df = filter_by_column_content(copy_df, "formtype", ["0001"])
+
+    # Filter for good statuses
+    fil_df = apply_filters(long_df)
+
+    # Fill zeros for no r&d and apply to original
+
+    df = clean_no_rd_data(fil_df, df, target_variables_list)
+
+    # Calculate imputation classes for each row
+    imp_df = create_imp_class_col(df, "200", "201", "imp_class")
+
+    return imp_df
+
+
+def sort(target_variable: str, df: pd.DataFrame) -> pd.DataFrame:
+    sort_list = [
+        f"{target_variable}",
+        "employees",
+        "reference",
+    ]
+    sorted_df = df.sort_values(
+        by=sort_list,
+        ascending=[True, False, True],
+    )
+    # sorted_df.reset_index(drop=True, inplace=True)
```

Comment:
Do we want to maintain index or reindex  - df.reset_index(drop=Tru, inplace=True)

Reply:
Are the index numbers used as a reference? If you uncommented this code, sorted_df would contain the sorted version of your DataFrame df, and its index would be reset, meaning we couldn't use the index as a reference for records.